### ISSUE TITLE
Repository Size Limit

### ISSUE THREAD

**Issue by jonasfranz** [go-gitea/gitea#3658]
It would be nice if it possible to set a global or user repository size limit. This would be good for Gitea providers providing Gitea for the public with limited disk space to have this option as mentioned in #1029 .

# Approaches for restricting repo size
## Solution 1
* User gets a notification if a repository exceeded the size limit 
* Admin gets a notification if a repository exceeded the size limit *(optional)*
* Repository gets deleted (after a ultimatum) if a user did not reduced the size of the repository

## Solution 2
* The size of every push gets approved (I don't know if that is possible or realistic) and blocked if the push is to large

**Comment by sbrl** [comment]
Suggestion: Have a warning limit too. That way users are warned when they are, say, 80% full - so they've got some time to do something about it.

**Comment by kolaente** [comment]
I'd prefer the second solution + warning. It doesn't seem practical to completely delete a users repo if it is past the size limit.

**Comment by jonasfranz** [comment]
@kolaente Do you have an idea who to calculate the size of a push before it is finished?

**Comment by sapk** [comment]
@JonasFranzDEV seems possible with a pre-receive hook that check size (cat-files) of the commit object. For examples : https://github.com/github/platform-samples/tree/master/pre-receive-hooks & https://stackoverflow.com/questions/40697663/show-commit-size-in-git-log

**Comment by jonasfranz** [comment]
@sapk Thank you for the hint. It seems that `size, err := git.GetRepoSize(repoPath)` will return the size of the repository including the newest push if it is called at pre-receive. 

How should the size limit work? 

1. Global size limit via the config file for all repositories
2. Custom size limit for repositories and a default value from the config file 
  a. How can change the limit of a repository and how? I propose the admin but I do not know if the admin panel or the repository settings are the right place to do this?
3. Size limit per user via config file with the option the change the size limit via the admin panel like it is possible for the number of repositories. (user administration)

Other proposals are welcome too!

**Comment by techknowlogick** [comment]
We could have a combination of the above, a global size limit to make sure that the disk doesn't run out of space, and a per user limit (in the case of some git hosts they say each user gets 1GB or something).

**Comment by jonasfranz** [comment]
@techknowlogick Should a push (no matter which user) be restricted/denied, if the global size limit is exceeded?

**Comment by lafriks** [comment]
I don't think 1st one is needed but would be great if it was combination of 2nd and 3rd. With defaults in config and custom values stored in repo and user tables with 0 being that default value limit should be applied and -1 that there is no limit

**Comment by lafriks** [comment]
Oh one more thing - should LFS be counted and if so than that should probably be done seperetly

**Comment by jonasfranz** [comment]
@lafriks 
> How can change the limit of a repository and how? I propose the admin but I do not know if the admin panel or the repository settings are the right place to do this?

**Comment by lafriks** [comment]
@JonasFranzDEV for user limit in user editing in admin panel. For repository there is now also a admin specific option in repo settings already

**Comment by sapk** [comment]
Also what happen when a user try to push to an org ? or did we consider org like a distinct user ?

**Comment by sapk** [comment]
Also what happen when a user try to push to an org ? or di we consider org like a distinct user ?

**Comment by lafriks** [comment]
I would say that org should be treated as separate user

**Comment by techknowlogick** [comment]
I don't think we should rely on just 2 & 3 as perhaps if you don't want to limit the number of users but you want to make sure you don't run out of space, then a global limit would be needed (ex. try.gitea.io runs out of space and if a user maxes out their space you don't want to encourage them to create an additional account). This is especially possible with an open gitea instance.

Regarding what we should do when the limit is reached, there are two options that I see:
1. show warnings in the git push prompt, and perhaps raise the warning is system messages of admin panel
2. reject push

or have two limits: soft (1), and hard (2)

**Comment by kolaente** [comment]
We should think of a way to prevent the following:

A user creates a repo, which means he is now admin of said repo. As a repo admin he'd be able to change the repo size limit and bypass the global  limit set by the server admin...

I guess the simplest solution would be to only allow to change the size limit setting if the user is also a server admin.

**Comment by lafriks** [comment]
@kolaente there is already a section in repo Settings that is available for server admin

**Comment by jonasfranz** [comment]
@techknowlogick I do not agree with that because I think that this would deny every push to gitea at all if the global limit is exceeded. So we need a way to restrict the size per user because this would not restrict users having only small repos from pushing.

@lafriks Would this be a posibility to get unlimited storage by creating organizations? My ideas:

a) Bind the limit of an organization to the owner of an organization
b) Accept @lafriks proposal and limit the count of organizations per user
c) Bind the limit of an organization to the members of the organization (might be complicated for users and developers due to the relationship between members and maximum size). Example: Organization uses 10GB of storage every user has 5GB of storage, The org has 4 members. => Every member could only use 2.5GB for his personal account because 2.5GB is used by the organization. (2.5GB*4=10GB)

**Comment by kolaente** [comment]
@JonasFranzDEV 
c) You mean when pushing to a repo inside that org or in general (aka when creating normal repos under the normal account)?

I'd go with @lafriks here, I think it would save us a lot of headache. 

Another thing: we should treat migrations like normal repos (in terms of limits), right? This would mean updating a migration should fail if a user has exceeded his limit. And we could check if a user has enough space left when creating the migration instead of doing that later on.

**Comment by lunny** [comment]
I think there are two different concepts here. One is repository size limit, another is user upload size limit. A repository size is the repository's folder size. User upload size is the sum size of all the commits he uploaded.

**Comment by jonasfranz** [comment]
@lunny it's quite easy to get the size of a repository after a push but I think that it would be harder to get the size of a commit itself. So I would propose to use repository size. The idea was to limit the size of a repository by a user limit.

**Comment by ghost** [comment]
@poVoq as an interim solution consider suggesting users run the [BFG repo cleaner](https://github.com/rtyley/bfg-repo-cleaner) to remove large files (especially video or large binaries) from their project history.

TBH disk space is cheap these days. I understand the word "cheap" is subjective but a little ingunity can go a long way. Let us not be held back by our tools but by our imaginations.

**Comment by yatsyk** [comment]
Is it true that any user with write permission to repository could disrupt the service for everyone by pushing quite big data to his personal repository?

**Comment by sbrl** [comment]
In theory yes, @yatsyk - as far as I understand.

Unless you're running a public instance though, it's unlikely to happen - and especially not on purpose.

**Comment by yatsyk** [comment]
@sbrl why do you think that it’s unlikely to happen?

**Comment by sbrl** [comment]
@yatsyk If you're _not_ running a public instance and an instance for you and some friends / co-workers, I would assume that you'd trust them enough not to intentionally try and break the server. If it does happen by accident, as others have said the BFG repo cleaner can help sort out the mess.

Of course, you will have unique requirements for your particular use-case.

**Comment by yatsyk** [comment]
@sbrl we should validate data in any service is it public or not. Co-worker computer could be hacked and we should not compromise other users.

**Comment by alexanderadam** [comment]
@sapk what will happen with mirrors? Will they be handled exactly the same?

**Comment by sapk** [comment]
@alexanderadam I think my PR is ignoring this case. I will to check if the pre-receive hook is triggered by mirror.

**Comment by jrtechs** [comment]
Is there any workaround for this right now? I'm getting the warning:
```
error: RPC failed; HTTP 413 curl 22 The requested URL returned error: 413 Request Entity Too Large
```

**Comment by sapk** [comment]
@jrtechs This is not implemented (yet). Your problem must be related to the proxy you used in front of gitea (like https://github.com/go-gitea/gitea/issues/5805).

**Comment by jrtechs** [comment]
@sapk it was. Thanks!

**Comment by peterbabic** [comment]
Please, is there any progress in the topic? I would like to prevent malicious users filling my drive - is there already some way?

**Comment by ClairelyClaire** [comment]
Will this be implemented? I'd like to use Gitea publicly but only allow users other than myself to file issues, not create repos. This will be difficult without a way to automatically set all new users to a repository limit of 0.

**Comment by lesderid** [comment]
Not being able to set a limit allowed a (presumably unintended) denial-of-service on my public instance.

**Comment by MrGeorgen** [comment]
any updates? otherwise it is not really possible to host it publicly

**Comment by danger89** [comment]
I really need this feature! 

My server is getting too busy with very very large git repos! Like 32GB of repos.

I would like to keep my server open for users, but 0,1% of the users screws-up for the rest of the users.

**Comment by mewalig** [comment]
We are implementing this ourselves as we cannot wait. If anyone would like to collaborate, pls lmk

**Comment by mewalig** [comment]
Thanks for pointing out, we did see and review that PR. However, since the PR code appears to never have been finished and also was made against a different version than what we are running, we felt we had to choose between "invest an uncertain amount of time/resources to determine whether that code would be worth attempting to reuse, with the best-case outcome still requiring further work to integrate into our customized-- and yet not latest-- gitea codebase" vs "invest a predictable amount of time/resources to implement this relatively small feature ourselves"-- and we went with the latter.

I would certainly have preferred to collaborate and spend those several-thousand bucks on something else, and furthermore I would have been happy to share the work we have done on this feature-- if only that could be done modular fashion that did not require a lot of additional work and cost for us. Unfortunately, from our perspective, those last two conditions are not met (I think this speaks to the enormous value, that is currently being lost, of having plug-in support as mentioned in #16195). Perhaps that is the fault of our own poor engineering decisions, but whatever the reason may be, our experience, real or perceived, is that while we would prefer to collaborate (both as a recipient of and donor of code for common features such as this) and we have budget to contribute, we are finding it difficult to do so in a way that makes economic sense.

Given now how increasingly far apart our gitea code bases are, I am increasingly of the view that the plug-in framework is a necessary precondition to enabling collaboration on other features.

**Comment by cGIfl300** [comment]
I know gitea is made to be self hosted but many don't have the knowledge to do it, why not sharing gitea? For that we need quotas limits, for the number of users and the volume of data they can use on disk. That would be a nice functionnality.

**Comment by ptman** [comment]
@cGIfl300 there's codeberg and gitea.com

**Comment by cGIfl300** [comment]
This is not what I mean, I mean each one could open an instance and allow a limited number of account, using a central system is not what I like in self hosting.

**Comment by mewalig** [comment]
Hi @DmitryFrolovTri we didn't create a PR because the version of gitea we are working from is now far behind the current version, and when we looked at the PR diff it just didn't make sense. I am still of the view that in the absence of the core maintainers agreeing to incorporate this into the core codebase (assuming the contributed code is of reasonable quality and does what it's supposed to do), it's unlikely to be useful to anyone to attempt to contribute this feature until and unless there is some sort of plug-in mechanism

**Comment by DmitryFrolovTri** [comment]
Well funny enough I also have a full PR for the old version that was done a while ago, it is very far behind.

**Comment by DmitryFrolovTri** [comment]
Sure. Please advise if any ideas that we are trying to implement here are contradicting

**Comment by DmitryFrolovTri** [comment]
I keep the list of things for this feature TO-DO in the
https://github.com/go-gitea/gitea/pull/21820

**Comment by DmitryFrolovTri** [comment]
We are moving ahead with this so hopefully we can have it done soon PR https://github.com/go-gitea/gitea/pull/21820

**Comment by melroy89** [comment]
> We are moving ahead with this so hopefully we can have it done soon PR #21820

What would be awesome for my public running server! And later git LFS. In my case without git LFS repo size limits, would be very welcome already.

**Comment by Cyberes** [comment]
Just chiming in with my support for this issue. I had a bot register on my Gitea instance and proceed to migrate multi-gig repositories from numerous git servers around the internet.

**Comment by ledlamp** [comment]
I can't believe this is not implemented, some bastard crashed my whole vps by uploading infinitely large repository...

<img width="1151" height="1027" alt="Image" src="https://github.com/user-attachments/assets/414d265c-1a89-4b09-b420-79fa05fb0e4f" />

**Comment by melroy89** [comment]
> I can't believe this is not implemented

I still can't believe it.

**Comment by DmitryFrolovTri** [comment]
> I can't believe this is not implemented, some bastard crashed my whole vps by uploading infinitely large repository...


Unfortunately it is true. I had no time to work on it and nobody suggested to take over. Managed to restore the condition so it will work with upstream (almost) a new merge conflict to fix. Which was just done (I hope)

**Comment by DmitryFrolovTri** [comment]
> I can't believe this is not implemented, some bastard crashed my whole vps by uploading infinitely large repository...

Hmm the current fix doesn't prevent this from happening I will add this to the list so we don't forget.

**Comment by lunny** [comment]
Before the pre-receive hook is executed, Git has already received the incoming packfile from the client and written it to disk as temporary files in the repository.

**Comment by DmitryFrolovTri** [comment]
> Before the pre-receive hook is executed, Git has already received the incoming packfile from the client and written it to disk as temporary files in the repository.

Yes so to prevent what happend to @ledlamp the size shall be enforced also during the package upload (server.go I presume) if total expected upload or already happened is over the limit such should be aborted. We assume it is packed and would contain even more data or this can be done as a separate item if you prefer, but I think it is better to use with the sizelimit

**Comment by hramrach** [comment]
The tmp packs are fun. I once set up push mirroring of the Linux sources to a forge. And it was over http, and they had some stupid proxy in front of the forge that cut the connection after some arbitrary timeout. The timeout was not long enough to upload the whole Linux sources. Over time the tmp packs took all 10T or how much storage they had for the forge.

**Comment by ledlamp** [comment]
FWIW, it appeared the repository that filled up all my space with tmp packs was a mirror of another large repository (11GiB) on my server, so maybe there is another issue with mirroring. Also it seemed none of the maintenance operations clean up the tmp packs so I'm not sure how they are supposed to be cleaned but I purged the user to free up the space.

Well, mirrors I can disable, but looks like there's nothing stopping anyone from abusing the core functionality and pushing a 1TB git repository..

**Comment by DmitryFrolovTri** [comment]
> FWIW, it appeared the repository that filled up all my space with tmp packs was a mirror of another large repository (11GiB)
 
Well unfortunately nothing that is being done at the moment is actually fixing this. mirror operations are not touched by any size limit. I will take a note on that.

**Comment by DmitryFrolovTri** [comment]
Got the https://github.com/go-gitea/gitea/pull/21820 to the state that it is working with Gitea development. Let's see if or how the review goes.

**Comment by hramrach** [comment]
Related to this it seems that only the site administrator can run garbage collection on repositories, not the repository owner.

Is there any particular reason to forbid garbage collection?

With size limits in place the repository owners would have an actual use case other than removing objects that contain leaked secrets or such.

**Comment by delvh** [comment]
Agreed.
The only reason why repo GC isn't implemented appears to be that no one considered adding it.

**Comment by DaanSelen** [comment]
Has there been updates on this?

**Comment by eyueldk** [comment]
This is honestly the only feature holding me back from using gitea on a project. Unable to control a user's quota means they can virtually DoS my entire service... sad.
